README.md
MoneyHelper [![CircleCI][ci_badge]][circleci]
============

A simple module to assist in formatting unambiguous prices and price ranges in international currencies in a Roman Script environment.

Usage
-----

Examples:

``` ruby
money_to_text(10000, currency: 'AUD')
#=> 'AUD $100.00'

money_to_text(10000, currency: 'AUD', with_currency: false, with_symbol: false)
#=> '100.00'

money_range_to_text(10000, 40000, currency: 'AUD')
#=> 'AUD $100.00 - 400.00'

money_range_to_text(10000, 40000, currency: 'AUD', delimiter: ' ... ')
#=> 'AUD $100.00 ... 400.00'
```

Defaults
--------

As of [#16][#16] this library explicitly specifies the `Money.locale_backend = :currency` default.

Contributing
------------

Fork the project. Make your feature addition or bug fix with tests. Send a pull request.

Copyright and License
---------------------

MIT License, see [LICENSE][license] for details.

(c) 2013-2018 [Artsy][artsy], [Sahil Yakhmi][sahil_yakhmi], [Joey Aghion][joey_aghion] and [contributors][contributors].

[ci_badge]: https://circleci.com/gh/artsy/money_helper/tree/main.svg?style=svg
[circleci]: https://circleci.com/gh/artsy/money_helper/tree/main
[#16]: https://github.com/artsy/money_helper/pull/16
[license]: LICENSE.md
[artsy]: https://github.com/artsy
[sahil_yakhmi]: https://github.com/syakhmi
[joey_aghion]: https://github.com/joeyAghion
[contributors]: CHANGELOG.md


CHANGELOG.md
### X.X.X (Next)

* Your contribution here.
### 3.0.2 (2023-03-22)

* [#30](https://github.com/artsy/money_helper/pull/30): Update `money_range_to_text` to receive `format` parameter for optional formatting - [@leamotta](https://github.com/leamotta).
### 3.0.1 (2023-03-21)

* [#29](https://github.com/artsy/money_helper/pull/29): Update `money_range_to_text` to receive `with_currency` parameter for optional ISO code printing - [@leamotta](https://github.com/leamotta).

### 3.0.0 (2022-02-16)

* [#27](https://github.com/artsy/money_helper/pull/27): Update `money_to_text` to pass `format` options to `Money#format` for additional features such as `no_cents` - [@agrberg](https://github.com/agrberg).

### 2.0.0 (2021-09-20)

#### Breaking Changes

* [#20](https://github.com/artsy/money_helper/pull/20): Update `money_to_text` and `money_range_to_text` to expect amounts in minor unit and simplify formatting; drop support for Ruby 2.3 and 2.4 - [@starsirius](https://github.com/starsirius)

### 1.0.2 (2019-04-19)

* [#16](https://github.com/artsy/money_helper/pull/16): Remove deprecation warnings - [@sweir27](https://github.com/sweir27).
* [#15](https://github.com/artsy/money_helper/pull/15): Added Rubocop - [@dblock](https://github.com/dblock).

### 1.0.1 (2017-04-03)

* Relaxed activesupport version dependency - [@dblock](https://github.com/dblock).
* The `money_to_text` method now takes options - [@mzikherman](https://github.com/mzikherman).
* Added `symbol_with_optional_iso_code` - [@mzikherman](https://github.com/mzikherman).

### 1.0.0 (2015-03-07)

* Added compatibility with Money 6.5.1+, including new currency symbols - [@syakhmi](https://github.com/syakhmi).

### 0.0.5 (2014-04-12)

* Added compatibility with Money 6.1+ - [@syakhmi](https://github.com/syakhmi).
* Dropped support for pre 6.0 versions of Money - [@syakhmi](https://github.com/syakhmi).

### 0.0.4 (2014-01-08)

* Added compatibility with Money 6.x - [@dblock](https://github.com/dblock).
* Avoid failing on deprecated currencies (such as ITL) - [@joeyAghion](https://github.com/joeyAghion).

### 0.0.3 (2013-07-17)

* Added support for currencies that don't have a symbol, eg. Uzbekistan Som (UZS) - [@dblock](https://github.com/dblock).

### 0.0.2 (2013-07-16)

* Added unit tests - [@syakhmi](https://github.com/syakhmi).
* Added support for Cambodian riel - [@syakhmi](https://github.com/syakhmi).
* Added dependencies to gemspec - [@syakhmi](https://github.com/syakhmi).
* Now defaults to USD when currency not specified - [@syakhmi](https://github.com/syakhmi).

### 0.0.1 (2013-07-12)

* Initial public release - [@syakhmi](https://github.com/syakhmi).


spec/spec_helper.rb
# frozen_string_literal: true

$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), '..'))

require 'rubygems'
require 'rspec'

require 'money_helper'


spec/version_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe MoneyHelper do
  it 'has a version' do
    expect(MoneyHelper::VERSION).to_not be nil
  end
end


spec/money_helper_spec.rb
# frozen_string_literal: true

require 'spec_helper'

describe MoneyHelper do
  describe 'money_to_text' do
    it 'includes ISO code and symbol' do
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AUD')).to eql('AUD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'ARS')).to eql('ARS $30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BBD')).to eql('BBD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BMD')).to eql('BMD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BND')).to eql('BND $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BSD')).to eql('BSD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BZD')).to eql('BZD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'CAD')).to eql('CAD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'CLP')).to eql('CLP $3.017.593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'COP')).to eql('COP $30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'CUC')).to eql('CUC $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'CUP')).to eql('CUP $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'DOP')).to eql('DOP $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'FJD')).to eql('FJD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'GYD')).to eql('GYD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'HKD')).to eql('HKD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'JMD')).to eql('JMD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'KYD')).to eql('KYD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'LRD')).to eql('LRD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'MXN')).to eql('MXN $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'NAD')).to eql('NAD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'NZD')).to eql('NZD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SBD')).to eql('SBD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SGD')).to eql('SGD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SRD')).to eql('SRD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'TWD')).to eql('TWD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'TTD')).to eql('TTD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'UYU')).to eql('UYU $U30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'USD')).to eql('USD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'XCD')).to eql('XCD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'ZWL')).to eql('ZWL $30,175.93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'EUR')).to eql('EUR €30.175,93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'FKP')).to eql('FKP £30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'GBP')).to eql('GBP £30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'GIP')).to eql('GIP £30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SDG')).to eql('SDG £30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SHP')).to eql('SHP £30,175.93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'CNY')).to eql('CNY ¥30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'JPY')).to eql('JPY ¥3,017,593')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'ALL')).to eql('ALL L30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'HNL')).to eql('HNL L30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'LSL')).to eql('LSL L30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'MDL')).to eql('MDL L30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'MYR')).to eql('MYR RM30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'RON')).to eql('RON Lei30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SZL')).to eql('SZL E30,175.93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'ANG')).to eql('ANG ƒ30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AWG')).to eql('AWG ƒ30,175.93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BWP')).to eql('BWP P30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'MOP')).to eql('MOP P30,175.93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'CRC')).to eql('CRC ₡30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SVC')).to eql('SVC ₡30,175.93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'MUR')).to eql('MUR ₨30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'NPR')).to eql('NPR Rs.30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'PKR')).to eql('PKR ₨30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SCR')).to eql('SCR ₨30,175.93')

      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AMD')).to eql('AMD դր.30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BDT')).to eql('BDT ৳30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BRL')).to eql('BRL R$30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'GEL')).to eql('GEL ლ30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'GHS')).to eql('GHS ₵30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'GMD')).to eql('GMD D30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'GTQ')).to eql('GTQ Q30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'HTG')).to eql('HTG G30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'IDR')).to eql('IDR Rp30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'ILS')).to eql('ILS ₪30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'INR')).to eql('INR ₹30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'KHR')).to eql('KHR ៛30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'KPW')).to eql('KPW ₩30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'LAK')).to eql('LAK ₭30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'MNT')).to eql('MNT ₮30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'NGN')).to eql('NGN ₦30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'NIO')).to eql('NIO C$30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'PGK')).to eql('PGK K30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'PHP')).to eql('PHP ₱30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'PYG')).to eql('PYG ₲3,017,593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'RUB')).to eql('RUB ₽30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'THB')).to eql('THB ฿30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'TMT')).to eql('TMT T30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'TOP')).to eql('TOP T$30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'TRY')).to eql('TRY ₺30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'UAH')).to eql('UAH ₴30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'VND')).to eql('VND ₫3.017.593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'WST')).to eql('WST T30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'ZAR')).to eql('ZAR R30,175.93')
    end

    it "respects currency's thousands separator and decimal mark" do
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'USD')).to eql('USD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'EUR')).to eql('EUR €30.175,93')
    end

    it "converts from currency's minor unit to major unit" do
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'USD')).to eql('USD $30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'JPY')).to eql('JPY ¥3,017,593')
    end

    it 'displays LTR for RTL symbols' do
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AFN')).to eql('AFN ؋30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'DZD')).to eql('DZD د.ج30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'BHD')).to eql('BHD د.ب3,017.593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'EGP')).to eql('EGP ج.م30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'IRR')).to eql('IRR ﷼30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'IQD')).to eql('IQD ع.د3,017.593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'JOD')).to eql('JOD د.ا3,017.593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'LYD')).to eql('LYD ل.د3,017.593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'MAD')).to eql('MAD د.م.30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'QAR')).to eql('QAR ر.ق30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'SYP')).to eql('SYP £S30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'TND')).to eql('TND د.ت3,017.593')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AED')).to eql('AED د.إ30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'YER')).to eql('YER ﷼30,175.93')
    end

    it 'defaults to USD' do
      expect(MoneyHelper.money_to_text(30_175_93)).to eql('USD $30,175.93')
    end

    it 'omits ISO code when with_currency is false' do
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'EUR', with_currency: false)).to eql('€30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AUD', with_currency: false)).to eql('$30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AMD', with_currency: false)).to eql('դր.30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AFN', with_currency: false)).to eql('؋30,175.93')
    end

    it 'omits symbol when `format: symbol:` is false' do
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'EUR', format: { symbol: false })).to eql('EUR 30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AUD', format: { symbol: false })).to eql('AUD 30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AMD', format: { symbol: false })).to eql('AMD 30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AFN', format: { symbol: false })).to eql('AFN 30,175.93')
    end

    it 'omits ISO code and symbol when both `with_currency` and `format: symbol:` are false' do
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'EUR', with_currency: false, format: { symbol: false })).to eql('30.175,93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AUD', with_currency: false, format: { symbol: false })).to eql('30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AMD', with_currency: false, format: { symbol: false })).to eql('30,175.93')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'AFN', with_currency: false, format: { symbol: false })).to eql('30,175.93')
    end

    it 'omits cents when `format: no_cents:` is true' do
      expect(MoneyHelper.money_to_text(30_175_93, with_currency: false, format: { no_cents: true, symbol: false })).to eql('30,175')
      expect(MoneyHelper.money_to_text(30_175_93, currency: 'EUR', with_currency: false, format: { no_cents: true, symbol: false })).to eql('30.175')
    end

    it 'returns an empty string if amount passed in is whitespace, empty string, or nil' do
      expect(MoneyHelper.money_to_text(' ')).to eql('')
      expect(MoneyHelper.money_to_text('')).to eql('')
      expect(MoneyHelper.money_to_text(nil)).to eql('')
    end

    it "raises an exception when currency can't be found" do
      expect do
        MoneyHelper.money_to_text(30_175_93, currency: 'AAA')
      end.to raise_error(Money::Currency::UnknownCurrency, "Unknown currency 'aaa'")
    end
  end

  describe 'symbol_with_optional_iso_code' do
    it 'just includes the symbol for USD GBP EUR and MYR' do
      expect(MoneyHelper.symbol_with_optional_iso_code('EUR')).to eql('€')
      expect(MoneyHelper.symbol_with_optional_iso_code('GBP')).to eql('£')
      expect(MoneyHelper.symbol_with_optional_iso_code('MYR')).to eql('RM')
      expect(MoneyHelper.symbol_with_optional_iso_code('USD')).to eql('$')
    end

    it 'includes the iso code as well for other currencies' do
      expect(MoneyHelper.symbol_with_optional_iso_code('AUD')).to eql('AUD $')
      expect(MoneyHelper.symbol_with_optional_iso_code('UZS')).to eql('UZS')
      expect(MoneyHelper.symbol_with_optional_iso_code('JPY')).to eql('JPY ¥')
    end
  end

  describe 'money_range_to_text' do
    it 'includes no indicator for currency for the upper amount in range' do
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'USD')).to eql('USD $30,175.93 - 40,983.27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AUD')).to eql('AUD $30,175.93 - 40,983.27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AMD')).to eql('AMD դր.30,175.93 - 40,983.27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AFN')).to eql('AFN ؋30,175.93 - 40,983.27')
    end

    it 'uses the special range amount delimeter when supplied' do
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'USD', delimiter: '-')).to eql('USD $30,175.93-40,983.27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AUD', delimiter: '-')).to eql('AUD $30,175.93-40,983.27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AMD', delimiter: '-')).to eql('AMD դր.30,175.93-40,983.27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AFN', delimiter: '-')).to eql('AFN ؋30,175.93-40,983.27')
    end

    it "prefixes the text 'Under ' when low amount not given" do
      expect(MoneyHelper.money_range_to_text(nil, 40_983_27, currency: 'USD')).to eql('Under USD $40,983.27')
    end

    it "prefixes the text 'Under ' when low amount not given and omits ISO code when with_currency is false" do
      expect(MoneyHelper.money_range_to_text(nil, 40_983_27, currency: 'USD', with_currency: false)).to eql('Under $40,983.27')
    end

    it "appends the text ' and up' when high amount not given" do
      expect(MoneyHelper.money_range_to_text(30_175_93, nil, currency: 'USD')).to eql('USD $30,175.93 and up')
    end

    it "appends the text ' and up' when high amount not given and omits ISO code when with_currency is false" do
      expect(MoneyHelper.money_range_to_text(30_175_93, nil, currency: 'USD', with_currency: false)).to eql('$30,175.93 and up')
    end

    it 'treats as a single price when low amount and high amount are identical' do
      expect(MoneyHelper.money_range_to_text(30_175_93, 30_175_93, currency: 'USD')).to eql('USD $30,175.93')
    end

    it 'treats as a single price when low amount and high amount are identical and omits ISO code when with_currency is false' do
      expect(MoneyHelper.money_range_to_text(30_175_93, 30_175_93, currency: 'USD', with_currency: false)).to eql('$30,175.93')
    end

    it 'returns empty string when both amounts are nil' do
      expect(MoneyHelper.money_range_to_text(nil, nil, currency: 'USD')).to eql('')
    end

    it 'omits ISO code when with_currency is false' do
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'EUR', with_currency: false)).to eql('€30.175,93 - 40.983,27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AUD', with_currency: false)).to eql('$30,175.93 - 40,983.27')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'AMD', with_currency: false)).to eql('դր.30,175.93 - 40,983.27')
    end

    it 'omits cents when `format: no_cents:` is true' do
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, with_currency: false, format: { no_cents: true })).to eql('$30,175 - 40,983')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'EUR', with_currency: false, format: { no_cents: true })).to eql('€30.175 - 40.983')
      expect(MoneyHelper.money_range_to_text(30_175_93, 30_175_93, currency: 'EUR', with_currency: false, format: { no_cents: true })).to eql('€30.175')
      expect(MoneyHelper.money_range_to_text(nil, 30_175_93, currency: 'EUR', with_currency: false, format: { no_cents: true })).to eql('Under €30.175')
      expect(MoneyHelper.money_range_to_text(30_175_93, nil, currency: 'EUR', with_currency: false, format: { no_cents: true })).to eql('€30.175 and up')
    end

    it 'omits cents when `format: no_cents:` is true, omits symbol when `format: symbol:` is false' do
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, with_currency: false, format: { no_cents: true, symbol: false })).to eql('30,175 - 40,983')
      expect(MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'EUR', with_currency: false, format: { no_cents: true, symbol: false })).to eql('30.175 - 40.983')
    end

    it "raises an exception when currency can't be found" do
      expect do
        MoneyHelper.money_range_to_text(10_000, 20_000, currency: 'ITL')
      end.to raise_error(Money::Currency::UnknownCurrency, "Unknown currency 'itl'")
    end
  end

  describe 'symbol_with_optional_iso_code' do
    it 'returns the symbol only if currency is in SYMBOL_ONLY list' do
      expect(MoneyHelper.symbol_with_optional_iso_code('EUR')).to eql('€')
      expect(MoneyHelper.symbol_with_optional_iso_code('USD')).to eql('$')
    end

    it 'returns iso code and symbol if symbol is in OK_SYMBOLS' do
      expect(MoneyHelper.symbol_with_optional_iso_code('INR')).to eql('INR ₹')
      expect(MoneyHelper.symbol_with_optional_iso_code('KHR')).to eql('KHR ៛')
      expect(MoneyHelper.symbol_with_optional_iso_code('KPW')).to eql('KPW ₩')
    end

    it 'returns only the iso code if symbol is not in OK_SYMBOLS' do
      expect(MoneyHelper.symbol_with_optional_iso_code('CHF')).to eql('CHF')
      expect(MoneyHelper.symbol_with_optional_iso_code('YER')).to eql('YER')
    end
  end
end


lib/money_helper.rb
# frozen_string_literal: true

require_relative 'version'

require 'active_support/core_ext/object/blank'
require 'money'

module MoneyHelper
  Money.locale_backend = :currency
  I18n.enforce_available_locales = false

  SYMBOL_ONLY = %w[USD GBP EUR MYR].freeze # don't use ISO code
  OK_SYMBOLS = %w[
    $ £ € ¥ 元 р. L ƒ ৳ P R$ K ₡ D ლ ₵ Q G ₹ Rp ₪ ₩ ₭ R RM ₨ ₮ դր. C$ ₦ ₲ ₱ T ฿ T$ m ₴ ₫ ៛ ₺ E ₽
  ].freeze # ok to include in string

  ##
  # Formats a single amount in the given currency into a price string. Defaults to USD if currency not
  #   given.
  #
  # = Example
  #
  #   MoneyHelper.money_to_text(30_175_93, currency: 'USD')  #=> 'USD $30,175.93'
  #   MoneyHelper.money_to_text(30_175_93, currency: 'EUR')  #=> 'EUR €30.175,93'
  #   MoneyHelper.money_to_text(30_175_93)                   #=> 'USD $30,175.93'
  #
  # = Arguments
  #
  #   amount_minor: (Integer) amount in minor unit
  #   currency: (String) optional ISO currency code, defaulting to USD
  #   with_currency: (Boolean) optional flag to include ISO currency code, defaulting to true
  # 　format: (Hash) optional formatting options to pass to `Money#format` e.g.:
  #     no_cents: (Boolean) optional flag to exclude cents, defaulting to false
  #     symbol: (Boolean) optional flag to include currency symbol, defaulting to true
  def self.money_to_text(amount_minor, currency: 'USD', with_currency: true, format: {})
    return '' if amount_minor.blank?

    format_options = { format: '%u%n' }.merge(format)

    formatted_amount = Money.new(amount_minor, currency).format(format_options)
    formatted_currency = with_currency ? currency.upcase : ''

    "#{formatted_currency} #{formatted_amount}".strip
  end

  def self.symbol_with_optional_iso_code(currency = 'USD')
    symbol = symbol_for_code(currency)
    if SYMBOL_ONLY.include?(currency)
      symbol
    elsif symbol && OK_SYMBOLS.include?(symbol)
      "#{iso_for_currency(currency)} #{symbol}"
    else
      iso_for_currency(currency).to_s
    end
  end

  ##
  # Formats a low and high amount in the given currency into a price string
  #
  # = Example
  #
  #   MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'USD')     #=> 'USD $30,175.93 - 40,983.27'
  #   MoneyHelper.money_range_to_text(30_175_93, 40_983_27)                      #=> 'USD $30,175.93 - 40,983.27'
  #   MoneyHelper.money_range_to_text(30_175_93, 40_983_27, currency: 'EUR')     #=> 'EUR €30.175,93 - 40.983,27'
  #   MoneyHelper.money_range_to_text(30_175_93, 40_983_27, delimiter: ' ... ')  #=> 'USD $30,175.93 ... 40,983.27'
  #
  # = Arguments
  #
  #   low: (Integer) amount in minor unit
  #   high: (Integer) amount in minor unit
  #   currency: (String) optional ISO currency code, defaulting to USD
  #   delimiter: (String) optional
  #   with_currency: (Boolean) optional flag to include ISO currency code, defaulting to true
  # 　format: (Hash) optional formatting options to pass to `Money#format` e.g.:
  #     no_cents: (Boolean) optional flag to exclude cents, defaulting to false
  #     symbol: (Boolean) optional flag to include currency symbol, defaulting to true
  def self.money_range_to_text(low, high, currency: 'USD', delimiter: ' - ', with_currency: true, format: {})
    if low.blank? && high.blank?
      ''
    elsif low.blank?
      "Under #{money_to_text(high, currency: currency, with_currency: with_currency, format: format)}"
    elsif high.blank?
      "#{money_to_text(low, currency: currency, with_currency: with_currency, format: format)} and up"
    elsif low == high
      money_to_text(low, currency: currency, with_currency: with_currency, format: format)
    else
      formatted_low = money_to_text(low, currency: currency, with_currency: with_currency, format: format)
      formatted_high = money_to_text(high, currency: currency, with_currency: false, format: { symbol: false }.merge(format))
      [formatted_low, formatted_high].compact.join(delimiter)
    end
  end

  def self.code_valid?(code)
    Money::Currency.stringified_keys.include?(code.downcase)
  end

  def self.iso_for_currency(code)
    return unless code && code_valid?(code)

    Money::Currency.new(code).iso_code.tap do |iso_code|
      iso_code.strip! if iso_code.present?
    end
  end

  def self.symbol_for_code(code)
    return unless code && code_valid?(code)

    Money::Currency.new(code).symbol.tap do |symbol|
      symbol.strip! if symbol.present?
    end
  end
end
